File: javascript/go-counting/go-counting.js
//
// This is only a SKELETON file for the 'Go Counting' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class GoCounting {
  constructor(board) {
    this.board = board;
    this.height = board.length;
    this.width = board[0].length;
  }

  getTerritory(x, y) {
    // Check if coordinates are valid
    if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
      return { error: 'Invalid coordinate' };
    }

    // Check if the position is a stone
    if (this.board[y][x] !== ' ') {
      return { owner: 'NONE', territory: [] };
    }

    // Find all connected empty intersections using flood fill
    const territory = [];
    const visited = new Set();
    const queue = [[x, y]];
    
    // Track neighboring stones to determine territory owner
    let blackNeighbors = false;
    let whiteNeighbors = false;

    while (queue.length > 0) {
      const [currentX, currentY] = queue.shift();
      const key = `${currentX},${currentY}`;
      
      if (visited.has(key)) continue;
      visited.add(key);

      // If it's an empty intersection, add to territory and queue its neighbors
      if (this.board[currentY][currentX] === ' ') {
        territory.push([currentX, currentY]);
        
        // Check all four directions (up, right, down, left)
        const directions = [
          [0, -1], [1, 0], [0, 1], [-1, 0]
        ];
        
        for (const [dx, dy] of directions) {
          const newX = currentX + dx;
          const newY = currentY + dy;
          
          // Skip if out of bounds
          if (newX < 0 || newX >= this.width || newY < 0 || newY >= this.height) {
            continue;
          }
          
          const cell = this.board[newY][newX];
          
          if (cell === ' ') {
            // Add empty intersection to queue
            queue.push([newX, newY]);
          } else if (cell === 'B') {
            // Found a black stone
            blackNeighbors = true;
          } else if (cell === 'W') {
            // Found a white stone
            whiteNeighbors = true;
          }
        }
      }
    }

    // Determine the owner of the territory
    let owner = 'NONE';
    if (blackNeighbors && !whiteNeighbors) {
      owner = 'BLACK';
    } else if (whiteNeighbors && !blackNeighbors) {
      owner = 'WHITE';
    }

    // Sort territory coordinates for consistent results
    territory.sort((a, b) => {
      if (a[0] !== b[0]) return a[0] - b[0];
      return a[1] - b[1];
    });

    return { owner, territory };
  }

  getTerritories() {
    const territoryBlack = [];
    const territoryWhite = [];
    const territoryNone = [];
    const visited = new Set();

    // Scan the entire board
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const key = `${x},${y}`;
        
        // Skip if already visited or not an empty intersection
        if (visited.has(key) || this.board[y][x] !== ' ') {
          continue;
        }

        // Get territory information for this position
        const { owner, territory } = this.getTerritory(x, y);
        
        // Add all coordinates in this territory to the visited set
        for (const [tx, ty] of territory) {
          visited.add(`${tx},${ty}`);
        }

        // Add territory to the appropriate category
        if (owner === 'BLACK') {
          territoryBlack.push(...territory);
        } else if (owner === 'WHITE') {
          territoryWhite.push(...territory);
        } else {
          territoryNone.push(...territory);
        }
      }
    }

    // Sort all territory coordinates for consistent results
    const sortCoordinates = (coords) => {
      return coords.sort((a, b) => {
        if (a[0] !== b[0]) return a[0] - b[0];
        return a[1] - b[1];
      });
    };

    return {
      territoryBlack: sortCoordinates(territoryBlack),
      territoryWhite: sortCoordinates(territoryWhite),
      territoryNone: sortCoordinates(territoryNone),
    };
  }
}

File: python/go-counting/go_counting.py
# Define constants for players
WHITE = "W"
BLACK = "B"
NONE = ""

class Board:
    """Count territories of each player in a Go game

    Args:
        board (list[str]): A two-dimensional Go board
    """

    def __init__(self, board):
        self.board = board
        self.height = len(board)
        self.width = len(board[0]) if self.height > 0 else 0
        # Cache for territories to avoid recalculating
        self._territories = None

    def territory(self, x, y):
        """Find the owner and the territories given a coordinate on
           the board

        Args:
            x (int): Column on the board
            y (int): Row on the board

        Returns:
            (str, set): A tuple, the first element being the owner
                        of that area.  One of "W", "B", "".  The
                        second being a set of coordinates, representing
                        the owner's territories.
        """
        # Check if coordinates are valid
        if not (0 <= x < self.width and 0 <= y < self.height):
            raise ValueError("Invalid coordinate")

        # If the position is not empty, it's not a territory
        if self.board[y][x] != " ":
            return NONE, set()

        # Use BFS to find all connected empty intersections
        territory = set()
        queue = [(x, y)]
        visited = {(x, y)}
        
        while queue:
            curr_x, curr_y = queue.pop(0)
            territory.add((curr_x, curr_y))
            
            # Check all four adjacent positions
            for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                nx, ny = curr_x + dx, curr_y + dy
                
                # Skip if out of bounds
                if not (0 <= nx < self.width and 0 <= ny < self.height):
                    continue
                
                # If empty and not visited, add to queue
                if self.board[ny][nx] == " " and (nx, ny) not in visited:
                    queue.append((nx, ny))
                    visited.add((nx, ny))
        
        # Determine the owner of the territory
        owner = self._determine_owner(territory)
        
        return owner, territory

    def _determine_owner(self, territory):
        """Determine the owner of a territory.
        
        A territory belongs to a player if all its borders are that player's stones.
        """
        borders = set()
        
        # Find all stones bordering the territory
        for x, y in territory:
            for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                nx, ny = x + dx, y + dy
                
                # Skip if out of bounds
                if not (0 <= nx < self.width and 0 <= ny < self.height):
                    continue
                
                # If it's a stone (not empty), add to borders
                if self.board[ny][nx] != " ":
                    borders.add(self.board[ny][nx])
        
        # If all borders are the same player, that player owns the territory
        if len(borders) == 1:
            return borders.pop()
        
        # If mixed or no borders, it's neutral
        return NONE

    def territories(self):
        """Find the owners and the territories of the whole board

        Args:
            none

        Returns:
            dict(str, set): A dictionary whose key being the owner
                        , i.e. "W", "B", "".  The value being a set
                        of coordinates owned by the owner.
        """
        # If we've already calculated territories, return the cached result
        if self._territories is not None:
            return self._territories
        
        # Initialize result dictionary
        result = {WHITE: set(), BLACK: set(), NONE: set()}
        
        # Keep track of visited positions
        visited = set()
        
        # Iterate through all positions on the board
        for y in range(self.height):
            for x in range(self.width):
                # Skip if not empty or already visited
                if self.board[y][x] != " " or (x, y) in visited:
                    continue
                
                # Find the territory and its owner
                owner, territory = self.territory(x, y)
                
                # Add to result and mark as visited
                result[owner].update(territory)
                visited.update(territory)
        
        # Cache the result
        self._territories = result
        
        return result
